index.rst

```diff
@@ -1,51 +1,1166 @@
+            xvec_c = np.array([-0.5, 0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5, 7.5, 8.5, 9.5])
+            yvec = np.array([1.0, 2.4, 5.0, 8.4, 13.0, 18.4, 25.0, 32.6, 41.0, 50.6])
+            yvec_c = np.array([1.0, 1.7, 3.7, 6.7, 10.7, 15.7, 21.7, 28.8, 36.8, 45.8, 50.6])
+
+            interp = afwMath.makeInterpolate(xvec, yvec, afwMath.Interpolate.CONSTANT)
+
+            for x, y in zip(xvec_c, yvec_c):
+                self.assertAlmostEqual(interp.interpolate(x + 0.1), y)
+                self.assertAlmostEqual(interp.interpolate(x), y)
+
+            self.assertEqual(interp.interpolate(xvec[0] - 10), yvec[0])
+            n = len(yvec)
+            self.assertEqual(interp.interpolate(xvec[n - 1] + 10), yvec[n - 1])
+
+            for x, y in reversed(list(zip(xvec_c, yvec_c))):  # test caching as we go backwards
+                self.assertAlmostEqual(interp.interpolate(x + 0.1), y)
+                self.assertAlmostEqual(interp.interpolate(x), y)
+
+            i = 2
+            for x in np.arange(xvec_c[i], xvec_c[i + 1], 10):
+                self.assertEqual(interp.interpolate(x), yvec_c[i])
+
+        #@unittest.skip("testing")
+        def testInvalidInputs(self):
+            """Test that invalid inputs cause an abort"""
+
+            self.assertRaises(pexExcept.OutOfRangeError,
+                              lambda: afwMath.makeInterpolate(np.array([], dtype=float), np.array([], dtype=float),
+                                                              afwMath.Interpolate.CONSTANT)
+                              )
+
+            afwMath.makeInterpolate(np.array([0], dtype=float), np.array([1], dtype=float),
+                                    afwMath.Interpolate.CONSTANT)
+
+            self.assertRaises(pexExcept.OutOfRangeError,
+                              lambda: afwMath.makeInterpolate(np.array([0], dtype=float), np.array([1], dtype=float),
+                                                              afwMath.Interpolate.LINEAR))
+
+
+    class TestMemory(lsst.utils.tests.MemoryTestCase):
+        pass
+
+    def setup_module(module):
+        lsst.utils.tests.init()
+
+    if __name__ == "__main__":
+        lsst.utils.tests.init()
+        unittest.main()
+
+Here we see that there is only one class called from this test: ``lsst::afw::math::Interpolate``. We make sure to add the appropriate lines to ``mathLib.py``, ``Sconscript``, and ``interpolate.cc`` as we saw in :ref:`new_cpp`.
+
+Smart Pointers
+^^^^^^^^^^^^^^
+
+We declare the class in the standard way, adding
+
+.. code-block:: c++
+
+    py::class_<Interpolate, std::shared_ptr<Interpolate>> clsInterpolate(mod, "Interpolate");
+
+to the module section of ``interpolate.cc``.
+``Interpolate`` itself is a virtual class and makes use of a shared pointer (see :ref:`virtual_functions`), so notice that here we added ``std::shared_ptr<Interpolate>`` as an inherited class, which is necessary to access Interpolate as a pointer.
+
```

Comment:
Somewhere (either here or "Frequently Encountered Problems") you should say that all related classes need to use the same holder type, and that the price of inconsistency is segfaults.